builtin/providers/azure/resource_azure_disk.go

```diff
@@ -0,0 +1,301 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "None",
+			},
+
+			"storage": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"media_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"source_media_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"virtual_machine": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceAzureDataDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+
+	if err := verifyDataDiskParameters(d); err != nil {
+		return err
+	}
+
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	p := virtualmachinedisk.CreateDataDiskParameters{
+		Lun:                 lun,
+		LogicalDiskSizeInGB: d.Get("size").(int),
+		HostCaching:         hostCaching(d),
+		MediaLink:           mediaLink(d),
+		SourceMediaLink:     d.Get("source_media_link").(string),
+	}
+
+	if name, ok := d.GetOk("name"); ok {
+		p.DiskName = name.(string)
+	}
+
+	log.Printf("[DEBUG] Adding data disk %d to instance: %s", lun, vm)
+	req, err := virtualmachinedisk.NewClient(mc).AddDataDisk(vm, vm, vm, p)
+	if err != nil {
+		return fmt.Errorf("Error adding data disk %d to instance %s: %s", lun, vm, err)
+	}
+
+	// Wait until the data disk is added
+	if err := mc.WaitForOperation(req, nil); err != nil {
+		return fmt.Errorf(
+			"Error waiting for data disk %d to be added to instance %s: %s", lun, vm, err)
+	}
+
+	log.Printf("[DEBUG] Retrieving data disk %d from instance %s", lun, vm)
+	disk, err := virtualmachinedisk.NewClient(mc).GetDataDisk(vm, vm, vm, lun)
+	if err != nil {
+		return fmt.Errorf("Error retrieving data disk %d from instance %s: %s", lun, vm, err)
+	}
+
+	d.SetId(disk.DiskName)
+
+	return resourceAzureDataDiskRead(d, meta)
+}
+
+func resourceAzureDataDiskRead(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	log.Printf("[DEBUG] Retrieving data disk: %s", d.Id())
+	datadisk, err := virtualmachinedisk.NewClient(mc).GetDataDisk(vm, vm, vm, lun)
+	if err != nil {
+		return fmt.Errorf("Error retrieving data disk %s: %s", d.Id(), err)
+	}
+
+	d.Set("name", datadisk.DiskName)
+	d.Set("lun", datadisk.Lun)
+	d.Set("size", datadisk.LogicalDiskSizeInGB)
+	d.Set("caching", datadisk.HostCaching)
+	d.Set("media_link", datadisk.MediaLink)
+
+	log.Printf("[DEBUG] Retrieving disk: %s", d.Id())
+	disk, err := virtualmachinedisk.NewClient(mc).GetDisk(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error retrieving disk %s: %s", d.Id(), err)
+	}
+
+	d.Set("virtual_machine", disk.AttachedTo.RoleName)
+
+	return nil
+}
+
+func resourceAzureDataDiskUpdate(d *schema.ResourceData, meta interface{}) error {
```

Comment:
This function seams to be a bit long. I wonder if its not better to abstract away some of the boilerplate that happens during most calls to the API. Looks like every call is 3 step:

* compose payload
* call API
* wait for result

A small wrapper around the azure API might help us de-duplicate some code and also make things easier to test.

So for example, we could have a type for virtual disk management that exposes some high level operations. And in this function we would do something similar to (pseudo code ahead):

```go
for {
    switch{
    case d.HasChange("size"):
        err := vmdisk.UpdateDisk(d)
    case d.HasChange("virtual_machine"):
        err := vmdisk.SetVirtualMachine(d)
    case d.HasChange("caching"), d.HasChange("lun"):
        err := vmdisk.UpdateDataDisk(d)
    default:
        return resourceAzureDataDiskRead(d, meta)
    }
}
```
This could be done for all other resources that duplicate code. These functions seem to be entry points for terraform. I believe it is best to remove as much of the business code from them into a separate type/interface/package as possible.  This way you decouple the implementation from the entry points, allowing you to easily switch it, if you need to, in the future. It also makes it easier for you to test and extend the functionality, without piling everything on top of a single function.

Thoughts?